### ISSUE TITLE
RSpec leaks ActiveSupport::ExecutionContext between specs.

### ISSUE THREAD

**Issue by BobbyMcWho** [rspec/rspec-rails#2644]
## What Ruby, Rails and RSpec versions are you using?

Ruby version: 2.7
Rails version: 7.0.4
RSpec version: ~> 6.0

## Observed behaviour

When using the Rails 7+ built in error reporter `Rails.error`, any context set during a spec leaks to the next spec.

I've resolved this by 
```ruby
config.before(:each)
  ActiveSupport::ExecutionContext.clear
end
```

While this works for me, it feels like a tripping point for folks using rspec-rails on rails >= 7.0

## Expected behaviour

ExecutionContext should be cleared between each spec, so that context does not leak between specs. 

## Can you provide an example app?

```ruby
require 'bundler/inline'

gemfile(true) do
  source 'https://rubygems.org'
  gem 'rails', '7.0.4'
  gem 'rspec-rails', '~> 6.0.0'
end

require 'rails'
require 'action_controller/railtie'

class MyApplication < Rails::Application; end

class PagesController < ActionController::Base
  include Rails.application.routes.url_helpers

  def index
    render inline: "SOME HTML"
  end
end

MyApplication.routes.draw do
  get '/pages', to: 'pages#index'
end

require 'rspec/rails'
require "rspec/autorun"

def app
  MyApplication
end

RSpec.describe "First non-controller test" do
  it "does not leak the controller context" do
    expect(ActiveSupport::ExecutionContext.to_h).to eq({})
  end
end

RSpec.describe PagesController, type: :controller do
  it "runs a controller test" do
    get :index
    expect(response.body).to eq("SOME HTML")
    expect(ActiveSupport::ExecutionContext.to_h).to eq({ controller: controller })
  end
end

# fails
RSpec.describe "Second non-controller test" do
  it "does not leak the controller context" do
    expect(ActiveSupport::ExecutionContext.to_h).to eq({})
  end
end

RSpec.describe "A context test" do
  before do
    ActiveSupport::ExecutionContext.clear
    Rails.error.set_context(foo: "bar")
  end

  it "sets the context" do
    expect(ActiveSupport::ExecutionContext.to_h).to eq({foo: "bar"})
  end
end

# fails
RSpec.describe "A second context test" do
  it "does not leak the context" do
    expect(ActiveSupport::ExecutionContext.to_h).to eq({})
  end
end
```

**Comment by pirj** [comment]
A quick solution would be to `include` [`ActiveSupport::ExecutionContext::TestHelper`](https://github.com/rails/rails/commit/6bad959565900499325d4239266cbb1f1f8e4056#diff-44565ad1bba1ae2e5ea44af9ed2beeb190fa2a466f2361fd0994b6040ce139a9R3) to [`RSpec::Rails::RailsExampleGroup`](https://github.com/rspec/rspec-rails/blob/c60ff7907559653cd9d1ec1a6113bf86c9359fab/lib/rspec/rails/example/rails_example_group.rb#L9).

What is interesting is that a new (can't find it in Rails 6.1 docs) option, [`executor_around_test_case`](https://guides.rubyonrails.org/configuring.html#config-active-support-executor-around-test-case), which defaults to `true` since 7.0, [basing on which](https://github.com/rails/rails/commit/6bad959565900499325d4239266cbb1f1f8e4056#diff-b4d3408d2c8d31897e3c9c761f414361dbdacb20556bb7b7a536fc3121de8c65R44), [`ActiveSupport::Executor::TestHelper`](https://github.com/rails/rails/blob/caaac482d5f430b29bb4c80e780fa70e65017755/activesupport/lib/active_support/executor/test_helper.rb#L3) would be included into [`ActiveSupport::TestCase`](https://github.com/rails/rails/blob/caaac482d5f430b29bb4c80e780fa70e65017755/activesupport/lib/active_support/test_case.rb#L21) instead.

Would you be interested in sending a PR, @BobbyMcWho ?

**Comment by javierjulio** [comment]
@pirj I've been reviewing this for awhile now since I was trying to understand if this would affect our Rails 7 upgrade and to see about contributing a fix/update. But I had been going around in circles trying to understand why this isn't supported and resetting `CurrentAttributes` is since they are treated the same. I recall some time ago coming across #2503 and with this comment https://github.com/rspec/rspec-rails/issues/2503#issuecomment-851054308 it details why some of the tests in this issue description are failing: they aren't specifying a type, e.g. `type: :model`. I was going to contribute a change but wonder now if this should be closed out instead, unless Rspec wants to support this and other Rails based helpers outside Rails-ish tests? Perhaps there is a special case here with the new `Rails.error` feature though.

**Comment by bensheldon** [comment]
I believe that the best option here is reproducing the behavior of [ActiveSupport::Executor::TestHelper](https://github.com/rails/rails/blob/caaac482d5f430b29bb4c80e780fa70e65017755/activesupport/lib/active_support/executor/test_helper.rb#L3) / [config.active_support.executor_around_test_case](https://guides.rubyonrails.org/configuring.html#config-active-support-executor-around-test-case) and wrapping every example by default with `Rails.application.executor.wrap {}`.  It's not _just_ ExecutionContext that isn't cleared, but the absence of a wrapping Executor can also alter Active Record behavior and other Rails behavior that expect to [always be wrapped within an Executor](https://guides.rubyonrails.org/threading_and_code_execution.html#wrapping-application-code).

**Comment by JonRowe** [comment]
`Rails.application.executor.perform {}` is not actually enough from testing but including the module if it exists works

**Comment by JonRowe** [comment]
If someone wants to take up the executor thing you can see it currently blows up various job things for us in #2712, given I don't really know what it affects and most internal rails things will have their own I'm inclined to leave it for now

**Comment by bensheldon** [comment]
@JonRowe I can take on the Executor wrap, thanks for linking to #2712. I'll make an issue to track it.

> I don't really know what it affects and most internal rails things will have their own I'm inclined to leave it for now

In Active Record, it will allow for the query cache to be enabled, it will also appropriately checkout and check-in database connections (and retry/recover from connection errors).

Also, the executor will only be present for integration-like tests because the only places Rails will inserts executors is via a Rack Middleware, or when deserializing a job via `perform_later`.

**Comment by bensheldon** [comment]
Created #2713

**Comment by JonRowe** [comment]
Fix technically released in 6.0.4 but please use 6.1.0 for Rails 7.1 support.
